modules/data.atmosphere/R/download.NOAA_GEFS_downscale.R

```diff
@@ -0,0 +1,345 @@
+##' 
+##' @section Information on Units:
+##' Information on NOAA weather units can be found below. Note that the temperature is measured in degrees C, but is converted at the station and downlaoded
+##' in Kelvin.
+##' @references https://www.ncdc.noaa.gov/crn/measurements.html
+##' 
+##' @section NOAA_GEFS General Information:
+##' This function downloads NOAA GEFS weather data.  GEFS is an ensemble of 21 different weather forecast models.  A 16 day forecast is avaliable
+##' every 6 hours.  Each forecast includes information on a total of 8 variables.  These are transformed from the NOAA standard to the internal PEcAn
+##' standard.
+##' 
+##' @section Data Avaliability:
+##' NOAA GEFS weather data is avaliable on a rolling 12 day basis; dates provided in "start_date" must be within this range. The end date can be any point after
+##' that, but if the end date is beyond 16 days, only 16 days worth of forecast are recorded.  Times are rounded down to the previous 6 hour forecast.  NOAA
+##' GEFS weather data isn't always posted immediately, and to compensate, this function adjusts requests made in the last two hours
+##' back two hours (approximately the amount of time it takes to post the data) to make sure the most current forecast is used.
+##' 
+##' @section Data Save Format:
+##' Data is saved in the netcdf format to the specified directory.  File names reflect the precision of the data to the given range of days.
+##' NOAA.GEFS.willow creek.3.2018-06-08T06:00.2018-06-24T06:00.nc specifies the forecast, using ensemble nubmer 3 at willow creek on
+##' June 6th, 2018 at 6:00 a.m. to June 24th, 2018 at 6:00 a.m.
+##' 
+##' @return A list of data frames is returned containing information about the data file that can be used to locate it later.  Each
+##' data frame contains information about one file.
+##'
+##' @param outfolder Directory where results should be written
+##' @param start_date, end_date Range of dates/times to be downloaded (default assumed time of day is 0:00, midnight)
+##' @param lat site latitude in decimal degrees
+##' @param lon site longitude in decimal degrees
+##' @param sitename The unique ID given to each site. This is used as part of the file name.
+##' @param overwrite logical. Download a fresh version even if a local file with the same name already exists?
+##' @param verbose logical.  Print additional debug information.  Passed on to functions in the netcdf4 package to provide debugging info.
+##' @param ... Other arguments, currently ignored
+##' @export
+##' 
+##' @examples 
+##' \dontrun{
+##'  download.NOAA_GEFS(outfolder="~/Working/results", lat.in= 45.805925, lon.in = -90.07961, sitename="US-WCr")
+##' }
+##' 
+##' @author Katie Zarada - modified code from Luke Dramko and Laura Puckett
+##' 
+##' 
+
+
+download.NOAA_GEFS_downscale <- function(outfolder, lat.in, lon.in, sitename, start_date = Sys.time(), end_date = (as.POSIXct(start_date, tz="UTC") + lubridate::days(16)),
```

Comment:
I've got a high-level design concern here.
1) A very large fraction of the code here is redundant with Luke's download.NOAA_GEFS, which is bad design as we'll have to make any changes twice
2) Conceptually with our met workflow, download and downscaling are two different steps and should be in two different modules. I'd highly recommend blowing away all of Luke's code, just retaining the downscaling code, and renaming it to make it clear that it's just a downscaling function. 

In doing so, I'd ask if anything in the downscaling is actually GEFS specific, or whether it should work with any other 6hr data (or data at other frequencies, such as 3hr). I'd also say that in pulling the downscaling out you'll want to make sure that works on netCDF CF files, not GEFS internal file formats.

Reply:
@mdietze This current version is very GEFS specific. I put the downscaling within the download function so that the downscaling occurs prior to it being written out as netCDF files. I was working with code from the FLARE group and so it was already GEFS specific and it seemed to be the easiest/make the most sense to do it that way. I can change it to be a generalized downscaling, but that will take some planning and time and is more complicated than the current set up. Where would you like that to fall on my list of priorities?

Reply:
@kzarada looking at the code, I don't actually think that there's anything in the downscaling that's actually particularly GEFS specific. All other forecast and reanalysis products that need downscaling will have the same variables and the same (or similar) timestep.

But the question about priorities is a good one, but not one that I can answer until I see the rest of the summer workplan you were going to sketch out after our Friday meeting.  Knowing this is lower priority than other tasks, for now I'm going to accept the PR, but do so knowing that this is a temporary, partial solution that requires refactoring.  FWIW, the refactoring itself should take < 1 hr, but what's going to be more time consuming is the insertion of the resultant downscaling function into the met workflow.